application/docker/Dockerfile

```diff
@@ -70,33 +61,128 @@ COPY --link ui/src/ src/
 ARG PUBLIC_API_BASE_URL=""
 ENV PUBLIC_API_BASE_URL ${PUBLIC_API_BASE_URL}
 
-#############
-# web_ui build
-#############
-FROM base_web_ui as web_ui
+########
+# Web UI
+########
+FROM web-ui-base as web-ui
 RUN npm run build
 
-#############
-# Server
-#############
-FROM python:3.13-slim@sha256:58c30f5bfaa718b5803a53393190b9c68bd517c44c6c94c1b6c8c172bcfad040 AS geti-tune
+######################
+# Server (CPU version)
+######################
+FROM python:3.13-slim@sha256:58c30f5bfaa718b5803a53393190b9c68bd517c44c6c94c1b6c8c172bcfad040 AS geti-tune-cpu
+
+WORKDIR /application
+ENV PYTHONPATH=/application
+
+# Install nginx & runtime system dependencies
+# TODO remove 'git' once all dependencies are available on PyPI
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+        git=1:2.47.3-0+deb13u1 \
+        libgl1=1.7.0-1+b2 \
+        libglib2.0-0=2.84.4-3~deb13u1 \
+        libglx-mesa0=25.0.7-2 \
+        nginx=1.26.* \
+    && rm -rf /var/lib/apt/lists/* \
+    && apt-get clean
 
-WORKDIR /app
-ENV PYTHONPATH=/app
+COPY docker/nginx.conf /etc/nginx/conf.d/default.conf
+COPY --from=web-ui /home/app/web_ui/dist/ /usr/share/nginx/html
+COPY --from=backend-base /application /application
+COPY --from=backend-base /bin/uv /bin/uv
 
-# Install nginx & OpenCV required libraries
-RUN apt-get update && apt-get install -y --no-install-recommends \
-    libgl1=1.7.0-1+b2 \
-    libglib2.0-0=2.84.4-3~deb13u1 \
-    libglx-mesa0=25.0.7-2 \
-    nginx=1.26.* \
+# Install Python dependencies for CPU version
+RUN --mount=type=cache,target=/root/.cache/uv \
+    --mount=type=bind,source=backend/uv.lock,target=uv.lock \
+    --mount=type=bind,source=backend/pyproject.toml,target=pyproject.toml \
+    uv sync --frozen --no-dev --no-editable --inexact --extra cpu
+
+EXPOSE 80
+
+CMD ["sh", "-c", "nginx && exec uv run ./app/main.py;"]
+
+######################
+# Server (GPU version)
+######################
+FROM python:3.13-slim@sha256:58c30f5bfaa718b5803a53393190b9c68bd517c44c6c94c1b6c8c172bcfad040 AS geti-tune-gpu
+
+WORKDIR /application
+ENV PYTHONPATH=/application
+
+RUN apt-get update  \
+    && apt-get install -y --no-install-recommends wget=1.25.0-2 \
+    && wget -q https://developer.download.nvidia.com/compute/cuda/repos/debian12/x86_64/cuda-keyring_1.1-1_all.deb \
+    && dpkg -i cuda-keyring_1.1-1_all.deb \
+    && rm -rf cuda-keyring_1.1-1_all.deb \
+    && apt-get remove -y wget
+
+# Install nginx & runtime system dependencies, including the Nvidia drivers.
+# Note that CUDA runtime dependencies are shipped through PyTorch, so there's no need to install the CUDA toolkit here.
```

Comment:
For reference: https://discuss.pytorch.org/t/what-is-the-interplay-between-the-nvidia-driver-and-cuda/217811/2